xtext-gradle-plugin/src/main/java/org/xtext/gradle/tasks/XtextGenerate.xtend

```diff
@@ -124,23 +125,22 @@ class XtextGenerate extends DefaultTask {
 		]
 	}
 
-	private def addIncrementalInputs(GradleBuildRequest request, IncrementalTaskInputs inputs) {
+	private def addIncrementalInputs(GradleBuildRequest request, InputChanges inputs) {
 		request.incremental = options.incremental && inputs.incremental
 
-		inputs.outOfDate[
-			if (allSources.contains(file)) {
-				request.dirtyFiles += file
+		for (change : inputs.getFileChanges(allSources)) {
```

Comment:
@oehme this does not work
```
Caused by: org.gradle.api.InvalidUserDataException: Cannot query incremental changes: No property found for value file collection. Incremental properties: mainSources.
	at org.gradle.internal.execution.history.changes.DefaultIncrementalInputProperties.getPropertyNameFor(DefaultIncrementalInputProperties.java:38)
	at org.gradle.internal.execution.history.changes.NonIncrementalInputChanges.getObjectFileChanges(NonIncrementalInputChanges.java:59)
	at org.gradle.internal.execution.history.changes.NonIncrementalInputChanges.getFileChanges(NonIncrementalInputChanges.java:50)
	at org.xtext.gradle.tasks.XtextGenerate.addIncrementalInputs(XtextGenerate.java:324)
	at org.xtext.gradle.tasks.XtextGenerate.generate(XtextGenerate.java:196)
	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:104)
	at org.gradle.api.internal.project.taskfactory.IncrementalInputsTaskAction.doExecute(IncrementalInputsTaskAction.java:32)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:51)
```
any hints?

Reply:
You need to annotate `getAllSources` as `@Incremental`, otherwise Gradle won't collect incremental change information on it. It's relatively expensive to do so, which is why it's deactivated by default.

Reply:
Caused by: org.gradle.api.InvalidUserDataException: Cannot query incremental changes: No property found for value file collection. Incremental properties: allSources
it looks like just a getter is not enhough

Reply:
It's recognizing the annotation, but it can't find the object in its map, because `sources.files` returns a different object on each call. A quick fix would be to cache that value after the first call [here](https://github.com/xtext/xtext-gradle-plugin/blob/4cab08d52646c6f941ba21d6b6f581739dea3b99/xtext-gradle-plugin/src/main/java/org/xtext/gradle/tasks/internal/DefaultXtextSourceDirectorySet.xtend#L52).